k2/csrc/fsa_algo.cc

```diff
@@ -0,0 +1,103 @@
+// k2/csrc/fsa_algo.cc
+
+// Copyright (c)  2020  Fangjun Kuang (csukuangfj@gmail.com)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/fsa_algo.h"
+
+#include <iostream>
+#include <sstream>
+#include <stack>
+
+namespace {
+
+// depth first search state
+struct DfsState {
+  int32_t s;      // state number of the visiting node
+  int32_t begin;  // arc index of the visiting arc
+  int32_t end;    // end of the arc index of the visiting node
+};
+
+}  // namespace
+
+namespace k2 {
+
+// The implementation of this function is inspired by
+// http://www.openfst.org/doxygen/fst/html/connect_8h_source.html
+void ConnectCore(const Fsa &fsa, std::vector<int32_t> *state_map) {
```

Comment:
@danpovey 
It seems the signature has to be changed.

There is no argument for the **out** fsa.